bundles/org.openhab.binding.solarman/src/main/java/org/openhab/binding/solarman/internal/SolarmanLoggerHandler.java

```diff
@@ -0,0 +1,222 @@
+                logger.info(
+                        "Assuming logger is OFFLINE after {} failed requests. Disabling connection error logging until it becomes available again",
+                        LoggerState.NO_FAILED_REQUESTS);
+            }
+        }, 0, config.refreshInterval, TimeUnit.SECONDS);
+    }
+
+    private <K, V> Map<K, V> mergeMaps(Map<K, V> map1, Map<K, V> map2) {
+        return Stream.concat(map1.entrySet().stream(), map2.entrySet().stream())
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, (v1, v2) -> v1));
+    }
+
+    private Map<ParameterItem, ChannelUID> extractChannelMappingFromChannels(List<Channel> channels) {
+        return channels.stream().map(channel -> {
+            BaseChannelConfig bcc = channel.getConfiguration().as(BaseChannelConfig.class);
+
+            @Nullable
+            String label = channel.getLabel();
+            if (label == null)
+                throw new IllegalStateException("Channel label should not be null");
+
+            return new AbstractMap.SimpleEntry<>(new ParameterItem(label, "N/A", "N/A", bcc.uom, bcc.scale, bcc.rule,
+                    parseRegisters(bcc.registers), "N/A", new Validation(), bcc.offset, Boolean.FALSE),
+                    channel.getUID());
+        }).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+    }
+
+    private List<Integer> parseRegisters(String registers) {
+        String[] tokens = registers.split(",");
+        Pattern pattern = Pattern.compile("\\s*(0x[\\da-fA-F]+|[\\d]+)\\s*");
+        return Stream.of(tokens).map(pattern::matcher).filter(Matcher::matches).map(matcher -> matcher.group(1))
+                .map(SolarmanLoggerHandler::parseNumber).collect(Collectors.toList());
+    }
+
+    // TODO for now just concatenate the list, in the future, merge overlapping requests
+    private List<Request> mergeRequests(List<Request> requestList1, List<Request> requestList2) {
+        return Stream.concat(requestList1.stream(), requestList2.stream()).collect(Collectors.toList());
+    }
+
+    private List<Request> extractAdditionalRequests(String channels) {
+        String[] tokens = channels.split(",");
+        Pattern pattern = Pattern.compile(
+                "\\s*(0x[\\da-fA-F]+|[\\d]+)\\s*:\\s*(0x[\\da-fA-F]+|[\\d]+)\\s*-\\s*(0x[\\da-fA-F]+|[\\d]+)\\s*");
+
+        return Stream.of(tokens).map(pattern::matcher).filter(Matcher::matches).map(matcher -> {
+            try {
+                int functionCode = parseNumber(matcher.group(1));
+                int start = parseNumber(matcher.group(2));
+                int end = parseNumber(matcher.group(3));
+                return new Request(functionCode, start, end);
+            } catch (NumberFormatException e) {
+                logger.error("Invalid number format in token: {} , ignoring additional requests", matcher.group(), e);
+                return (@NonNull Request) null;
```

Comment:
`@NonNull` should not be needed as the class is already annotated with `@NonNullByDefault`